Main svc

GreNaz opened the pull request.

Commit:
add main module structure

Commit:
add main module models

Commit:
add main module models +doc

Commit:
add main module Compilations API

Commit:
add main module Categories API

Commit:
add main module usersAdmin API

Commit:
add main module RequestsPrivate API

Commit:
main module event API frame adding

Commit:
main module event API EventsPrivate adding

Commit:
main module event API для работы с событиями

Commit:
main module Full API

Commit:
refactor structure

GreNaz committed:
Fix RqController -> RqMapping

GreNaz committed:
set User

Commit:
Main service add working User logic

Commit:
Main service add working User logic + Refactoring

Commit:
Main service add working Categories frame

GreNaz committed:
main service add Categories logic
refactoring user return code after delete object

GreNaz committed:
main service add Categories logic refactoring

GreNaz committed:
main service add Compilations add interface controller

GreNaz committed:
main service add API для работы с событиями (interface controller)

GreNaz committed:
main service add Закрытый API для работы с запросами текущего пользователя на участие в событиях
 (interface controller)

GreNaz committed:
main service add Закрытый API для работы с запросами текущего пользователя на участие в событиях
 (interface controller)

GreNaz committed:
Merge remote-tracking branch 'origin/main_svc' into main_svc

GreNaz committed:
main service  full reformat module code

GreNaz committed:
main service reformat code

GreNaz committed:
main service event entity

GreNaz committed:
main service add repository interface

GreNaz committed:
main service add full service classes frame

GreNaz committed:
main service inject services

GreNaz committed:
main service finish structure writing

GreNaz committed:
main service finish structure writing

GreNaz committed:
main service POST Event

GreNaz committed:
main service  Event

GreNaz committed:
main service  Request

GreNaz committed:
main service  Events get all

GreNaz committed:
main service  Events getEvents by User

GreNaz committed:
main service  Events getEvents Получение событий с возможностью фильтрации

GreNaz committed:
main service  Events getEvents Получение подробной информации об опубликованном событии по его идентификатору

GreNaz committed:
main service  Events getEvents Получение полной информации о событии добавленном текущим пользователем

GreNaz committed:
main service  Events getEvents Получение информации о заявках текущего пользователя на участие в чужих событиях

GreNaz committed:
main service  Events getEvents Получение информации о запросах на участие в событии текущего пользователя

GreNaz committed:
main service  Events getEvents Редактирование данных события и его статуса (отклонение/публикация). - ТРЕБУЕТСЯ ОТЛАДКА ПАДАЕТ ПРИ ПАРСИНГЕ ДАТЫ В СЕРВИСЕ

GreNaz committed:
main service  Events getEvents Редактирование данных события и его статуса (отклонение/публикация). DONE

GreNaz committed:
main service  Events getEvents Отмена своего запроса на участие в событии

GreNaz committed:
main service  Events getEvents Изменение статуса (подтверждена, отменена) заявок на участие в событии текущего пользователя

GreNaz committed:
main service  Compilations DONE

GreNaz committed:
main service  Validation Errors

GreNaz committed:
main service  Попытка изменения имени категории на уже существующее

GreNaz committed:
main service  Изменение события добавленного текущим пользователем. Изменение даты не неподходящую

GreNaz committed:
main service  409 conflict resolved

GreNaz committed:
reformat code

GreNaz committed:
ErrorHandler refactoring

GreNaz committed:
All general tests - ok

GreNaz committed:
Осталось отладить статистику

GreNaz committed:
Statistics - Done

GreNaz committed:
fix bugs

GreNaz committed:
fix bugs 45 ERRORS

GreNaz committed:
fix bugs 42 ERRORS

GreNaz committed:
fix bugs 33 ERRORS

GreNaz committed:
fix bugs 21 ERRORS

GreNaz committed:
fix bugs 19 ERRORS

GreNaz committed:
fix bugs 17 ERRORS

GreNaz committed:
fix bugs 11 ERRORS

GreNaz committed:
fix bugs 9 ERRORS

GreNaz committed:
fix bugs 7 ERRORS

GreNaz committed:
fix bugs 1 ERROR

GreNaz committed:
DONE

GreNaz committed:
DONE +

GreNaz committed:
DONE +

GreNaz committed:
DONE +

GreNaz committed:
DONE

GreNaz committed:
DONE

GreNaz committed:
DONE

GreNaz committed:
DONE ++

GreNaz committed:
DONE ++

GreNaz committed:
DONE ++

GreNaz committed:
DONE ++

GreNaz committed:
DONE ++

GreNaz committed:
DONE ++

GreNaz committed:
точно работает

GreNaz committed:
точно работает 2

TyutterinYakov commented on ewm-main/pom.xml:
```diff
@@ -45,7 +54,20 @@
+        <dependency>
+            <groupId>org.mapstruct</groupId>
+            <artifactId>mapstruct</artifactId>
+            <version>1.5.3.Final</version>
```
Хорошо было бы вынести версии зависимостей одной группы в блок `properties`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/controller/errors/ErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
```
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Throwable`, а возвращать статус 500 INTERNAL_SERVER_ERROR
Так же следует добавить обработчик для исключения, которое выбрасывается при валидации данных с помощью аннотаций - `MethodArgumentNotValidException` и `ConstraintViolationException`, возвращать будет статус 400 BAD_REQUEST, так как ошибка связана с некорректными данными от пользователя)

GreNaz replied:
Throwable - ломает все может я чего не понимаю 500 и так будут выбрасываться же?
По остальным добавил

TyutterinYakov replied:
Да, это стандартное поведение, так как добавление обработчики не для всех исключений)

GreNaz replied:
Ну так а что делать то с этим?

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/controller/errors/ErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.CONFLICT)
+    public ErrorResponse handleIntegrityException(final ConflictException e) {
```
Можно было бы не дублировать хендлеры, а описать похожие в одном, используя `@ExceptionHandler({@ExceptionHandler.class, ....})`
В качестве параметра метода будет выступать ближайший родитель) В данном случае - это `RuntimeException`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/controller/compilations/impl/CompilationsPublicControllerImpl.java:
```diff
@@ -0,0 +1,31 @@
+    private final CompilationService service;
+
+    public List<CompilationDto> getCompilations(Boolean pinned, Integer from, Integer size) {
+        if (pinned == null) {
```
По спецификации это значение не является обязательным) Если оно не было передано - следует произвести получение без учета `pinned`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/controller/users/UsersAdminController.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    @GetMapping()
+    List<UserDto> getUsers(@RequestParam(required = false) List<Long> ids,
+                           @RequestParam(defaultValue = "0") Integer from,
```
(предложение) Почему бы не использовать более декларативный подход и использовать аннотации ограничения) Например, для from можно было бы использовать `@PositiveOrZero`, так как значение может быть 0, либо больше) Аналогично и для size, но в этом случае следует использовать `@Positive`, так как значение не может быть равно 0

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/controller/users/impl/UsersAdminControllerImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+    public List<UserDto> getUsers(List<Long> ids, Integer from, Integer size) {
+        Pageable pageable = PageRequest.of(from, size);
+        if (ids == null) {
```
Если параметр `ids` не передан - получение должно быть выполнено без его учета) Инициализация не требуется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/categories/CategoryMapper.java:
```diff
@@ -0,0 +1,23 @@
+import ru.practicum.ewm.model.categories.dto.CategoryDto;
+import ru.practicum.ewm.model.categories.dto.NewCategoryDto;
+
+@UtilityClass
```
При использовании этой аннотации даже `static` указывать не требуется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/compilations/CompilationMapper.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+@UtilityClass
```
При использовании этой аннотации даже `static` указывать не требуется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/users/UserMapper.java:
```diff
@@ -0,0 +1,39 @@
+import ru.practicum.ewm.model.users.dto.UserDto;
+import ru.practicum.ewm.model.users.dto.UserShortDto;
+
+@UtilityClass
```
При использовании этой аннотации даже `static` указывать не требуется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/categories/CategoriesRepository.java:
```diff
@@ -0,0 +1,10 @@
+import ru.practicum.ewm.model.categories.Category;
+
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от `JpaRepository` мы уже объявляем бин)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/compilations/CompilationsRepository.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.List;
+
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от `JpaRepository` мы уже объявляем бин)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/EventsRepository.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Optional;
+
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от `JpaRepository` мы уже объявляем бин)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/location/LocationRepository.java:
```diff
@@ -0,0 +1,10 @@
+import ru.practicum.ewm.model.Location;
+
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от `JpaRepository` мы уже объявляем бин)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/requests/RequestsRepository.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Optional;
+
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от `JpaRepository` мы уже объявляем бин)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/users/UsersRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.List;
+
+@Repository
```
Аннотация `@Repository` не требуется, так как наследуясь от `JpaRepository` мы уже объявляем бин)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/categories/CategoriesService.java:
```diff
@@ -0,0 +1,25 @@
+import javax.transaction.Transactional;
+import java.util.List;
+
+@Service
```
Действительно ли здесь требуется эта аннотация?)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/categories/impl/CategoriesServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+    public CategoryDto updateCategory(CategoryDto categoryDto, Integer catId) {
+        Category category = checkCategoryExist(catId);
+        if (categoryDto.getName().equals(category.getName())) {
+//            throw new ConflictException("Same category name");
```
Если код не используется - следует его исключить)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/compilations/CompilationService.java:
```diff
@@ -0,0 +1,26 @@
+import javax.transaction.Transactional;
+import java.util.List;
+
+@Service
```
Действительно ли здесь требуется эта аннотация?)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/events/EventService.java:
```diff
@@ -0,0 +1,25 @@
+import javax.servlet.http.HttpServletRequest;
+import java.util.List;
+
+@Service
```
Действительно ли здесь требуется эта аннотация?)

TyutterinYakov commented on ewm-stat/stat-client/src/main/java/ru/practicum/ewm/client/StatClient.java:
```diff
@@ -0,0 +1,69 @@
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
+
+    public void save(EndpointHitDTO endpointHitDTO) {
+        log.error("Сохранение запроса на " + endpointHitDTO);
```
При логгировании более эффективно использовать не конкатенацию, а заполнитель - placeholder `{}`
Например:
`log.info("Я не найден. Мой id = {}", id);`
Аналогично в других местах)
Также использование уровня `error` - не совсем корректно, так как здесь записывается информация, достаточно было бы использовать `.info`

TyutterinYakov commented on ewm-stat/stat-client/src/main/java/ru/practicum/ewm/client/StatClient.java:
```diff
@@ -0,0 +1,69 @@
+                                          LocalDateTime end,
+                                          List<String> uris,
+                                          boolean unique) {
+        log.error("Получение статистики запроса(ов) " + uris);
```
При логгировании более эффективно использовать не конкатенацию, а заполнитель - placeholder `{}`
Например:
`log.info("Я не найден. Мой id = {}", id);`
Аналогично в других местах)
Также использование уровня `error` - не совсем корректно, так как здесь записывается информация, достаточно было бы использовать `.info`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/users/impl/UsersServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public void deleteUser(Long userId) {
+        //проверить наличие пользователя?
```
да, можно, иначе может быть выброшено `EmptyResultDataAccessException`)
Например:
```
if (existsById(userId)) {
     repository.deleteById(userId);
}
```
Используем `exists`, чтобы не производить лишнюю десериализацию) Но можно объявить свой метод `deleteById`, который не будет выбрасывать исключение)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/users/impl/UsersServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public List<UserDto> getUsers(List<Long> ids, Pageable pageable) {
+        if (!ids.isEmpty()) {
```
Теперь достаточно будет проверить, что значение не `null`)

TyutterinYakov commented on ewm-stat/stat-server/src/main/java/ru/practucum/ewm/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+@Slf4j
+@RestControllerAdvice
+public class ErrorHandler {
```
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Throwable`, а возвращать статус 500 INTERNAL_SERVER_ERROR

TyutterinYakov replied:
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Throwable`, а возвращать статус 500 INTERNAL_SERVER_ERROR
Так как мы должны ловить и обрабатывать все исключения, которые могут возникнуть при выполнении - соответствующим образом

TyutterinYakov commented on ewm-stat/stat-server/src/main/java/ru/practucum/ewm/repository/StatRepository.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+@Repository
```
Аннотация также не требуется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/compilations/impl/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+        if (updateCompilationRequest.getPinned() != null) {
+            compilation.setPinned(updateCompilationRequest.getPinned());
+        }
+        if (updateCompilationRequest.getTitle() != null) {
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)
Иначе при обновлении допустимо передать строку только из пробелов или пустую, что при создании недопустимо)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/compilations/impl/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public CompilationDto updateCompilation(UpdateCompilationRequest updateCompilationRequest, Integer compId) {
+        Compilation compilation = compilationsRepository.findById(compId).orElseThrow();
```
Лучше всегда указывать исключение, которое мы хотим выбросить и использовать то, что относится к этому виду ошибки) Чтобы из сообщения было проще понять, что именно пошло не так)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/compilations/impl/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public CompilationDto setCompilation(NewCompilationDto newCompilationDto) {
+        List<Event> events = eventsRepository.findByIds(newCompilationDto.getEvents());
```
Для начала следует проверить, что значение `newCompilationDto.getEvents()` не `null`, так как подборка может быть создана пустой)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/users/dto/UserShortDto.java:
```diff
@@ -0,0 +1,22 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserShortDto {
```
Действительно ли в этом классе требуются аннотации валидации?) Если я не ошибаюсь, то он выступает только в качестве возвращаемого значения, и оно никак не валидируется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/users/dto/UserDto.java:
```diff
@@ -0,0 +1,24 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserDto {
```
Действительно ли в этом классе требуются аннотации валидации?) Если я не ошибаюсь, то он выступает только в качестве возвращаемого значения, и оно никак не валидируется)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/users/dto/NewUserRequest.java:
```diff
@@ -0,0 +1,28 @@
+@NoArgsConstructor
+public class NewUserRequest {
+    @NotBlank
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в `@NotBlank`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/requests/RequestsMapper.java:
```diff
@@ -0,0 +1,15 @@
+
+@Mapper
+public interface RequestsMapper {
+    RequestsMapper REQUESTS_MAPPER = Mappers.getMapper(RequestsMapper.class);
```
(предложение)Можно было бы не объявлять дополнительно поле, а сделать наш мапер - бином)
https://www.baeldung.com/mapstruct
Иначе здесь нарушается принцип единственной ответственности)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/Location.java:
```diff
@@ -0,0 +1,27 @@
+@AllArgsConstructor
+@Entity
+@Table(name = "LOCATIONS")
+public class Location {
```
(предложение)Действительно ли требуется представлять Location в виде отдельной сущности БД?) Можно было бы над классом Location навесить аннотацию `@Embeddable`, таким образом при формировании схемы БД поля этого класса будут отображены как часть таблицы `events`) Это в разы экономнее, чем при извлечении событий дополнительно производить присоединение таблицы `locations`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/compilations/Compilation.java:
```diff
@@ -0,0 +1,30 @@
+
+    @ManyToMany
+    @JoinColumn(name = "events_id")
+    private List<Event> events;
```
В данном случае хорошо было бы использовать `Set`, так как события в подборке - уникальны)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/Event.java:
```diff
@@ -0,0 +1,68 @@
+    @Column(nullable = false, length = 500)
+    private String title;
+
+    private Integer views;
```
Хранение вычисляемых полей в БД - не является хорошей практикой. Хорошо было бы исключить это поле из таблицы БД

Также в ТЗ явно прописано, что просмотры должны храниться в сервисе статистики)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/compilations/dto/NewCompilationDto.java:
```diff
@@ -0,0 +1,26 @@
+
+    private List<Long> events; //Список идентификаторов событий входящих в подборку
+
+    private Boolean pinned; //Закреплена ли подборка на главной странице сайта
```
Хорошо было бы сделать тип - примитивом, так как значение по спецификации не является обязательным, а по дефолту должно быть не `null`, а `false`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/EventFullDto.java:
```diff
@@ -0,0 +1,39 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate; //Дата и время на которые намечено событие (в формате "yyyy-MM-dd HH:mm:ss")
+    private UserShortDto initiator;
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/EventRequestStatusUpdateRequest.java:
```diff
@@ -0,0 +1,18 @@
+@NoArgsConstructor
+public class EventRequestStatusUpdateRequest {
+    private List<Long> requestIds; //Идентификаторы запросов на участие в событии текущего пользователя
+    private String status; //Новый статус запроса на участие в событии текущего пользователя
```
В данном случае хорошо было бы использовать не `String`, а `Enum`, чтобы заранее производить проверку а соответствие допустимых статусов)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/NewEventDto.java:
```diff
@@ -0,0 +1,43 @@
+    @Size(min = 3, max = 120)
+    private String title; //Заголовок события
+
+    private Boolean paid; //Нужно ли оплачивать участие в событии
```
`paid` может иметь лишь два состояния) Хорошо было бы в данном случае использовать `@NotNull`, либо же изменить тип на примитив) Если значение не будет передано, в таком случае оно будет проинициализировано не null, а false)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/NewEventDto.java:
```diff
@@ -0,0 +1,43 @@
+    private Boolean paid; //Нужно ли оплачивать участие в событии
+    @PositiveOrZero
+    private Integer participantLimit; //Ограничение на количество участников. Значение 0 - означает отсутствие ограничения
+    private Boolean requestModeration; //Нужна ли пре-модерация заявок на участие. Если true, то все заявки будут ожидать подтверждения инициатором события. Если false - то будут подтверждаться автоматически.
```
Аналогично, `requestModeration` может иметь лишь два состояния
По спецификации стандартным значением является `true`, можно было бы сразу же произвести инициализацию)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/NewEventDto.java:
```diff
@@ -0,0 +1,43 @@
+
+    private Boolean paid; //Нужно ли оплачивать участие в событии
+    @PositiveOrZero
+    private Integer participantLimit; //Ограничение на количество участников. Значение 0 - означает отсутствие ограничения
```
Также хорошо было бы подправить тип на примитивный, так как `@PositiveOrZero` допускает `null` за корректное значение)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/NewEventDto.java:
```diff
@@ -0,0 +1,43 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate; //Дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    @NotNull
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться
Также можно было бы в будущем классе `dto` навесить над соответствующими полями аннотации валидации, а здесь - `@Valid`, чтобы они стали учитываться) Так как сейчас допустимо передать пустой `json`

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventAdminRequest.java:
```diff
@@ -0,0 +1,37 @@
+    private String description; //  Новое описание
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate; //  Новые дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventAdminRequest.java:
```diff
@@ -0,0 +1,37 @@
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
+public class UpdateEventAdminRequest {
```
(предложение)Можно было бы для классов обновления ввести общего родителя, чтобы не дублировать поля в двух классах)
У него будет два наследника: `UpdateEventUserRequest` и  `UpdateEventAdminRequest`, в каждом из которых будет объявлено поле: `stateAction`
Например для пользователя:
```
public class UpdateEventUserRequest extends UpdateEventRequest {
	private StateAction stateAction;
	public enum StateAction {
		SEND_TO_REVIEW,
		CANCEL_REVIEW
	}
}
```
Таким же образом и для администратора, но у него будут содержаться другие значения в `StateAction`
Таким образом мы сразу же будем проверять корректность поступающих данных и исключим дублирование)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventAdminRequest.java:
```diff
@@ -0,0 +1,37 @@
+    private LocalDateTime eventDate; //  Новые дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    private Location location;
+    private Boolean paid; //  Новое значение флага о платности мероприятия
+    private Integer participantLimit; //  Новый лимит пользователей
```
`participantLimit` может быть только положительным число, либо же 0. В данном случае было бы удобно использовать `@PositiveOrZero`, так как она удобна тем, что допускает null, по которому при обновлении мы сможем понять, было передано значение или нет

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventUserRequest.java:
```diff
@@ -0,0 +1,37 @@
+    private LocalDateTime eventDate; //  Новые дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    private Location location;
+    private Boolean paid; //  Новое значение флага о платности мероприятия
+    private Integer participantLimit; //  Новый лимит пользователей
```
`participantLimit` может быть только положительным число, либо же 0. В данном случае было бы удобно использовать `@PositiveOrZero`, так как она удобна тем, что допускает null, по которому при обновлении мы сможем понять, было передано значение или нет

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventUserRequest.java:
```diff
@@ -0,0 +1,37 @@
+    private String description; //  Новое описание
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate; //  Новые дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+    public static void getConfirmedRequests(List<EventFullDto> eventDtos,
+                                            RequestsRepository requestsRepository) {
+        List<Long> ids = eventDtos.stream().map(EventFullDto::getId).collect(Collectors.toList());
+        List<ParticipationRequest> requests = requestsRepository.findConfirmedToListEvents(ids);
```
(предложение) Более эффективно было бы воспользоваться механизмом проекций. 
Объявить класс с полями `count` и `eventId`, чтобы получить коллекцию из репозитория(да, можно получить ее и в виде `List< ParticipationRequest>` но это не очень эффективно, так как нам требуется лишь количество запросов), далее с помощью стримов ее было бы удобно преобразовать к мапе `Map<Long, Long>`, где ключ - идентификатор события, значение - количество подтвержденных запросов) Мапа нужна, чтобы получать количество запросов для события за O(1)
https://habr.com/ru/companies/otus/articles/722060/

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+                -> "/events/" + fullEventDto.getId(), fullEventDto -> fullEventDto));
+        Object responseBody = statService.getViewStats(
+                        MIN_TIME,
+                        MAX_TIME,
```
Дата окончания - это текущая дата, так как позже нее не может быть просмотров)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+        Map<String, EventFullDto> views = eventDtos.stream().collect(Collectors.toMap(fullEventDto
+                -> "/events/" + fullEventDto.getId(), fullEventDto -> fullEventDto));
+        Object responseBody = statService.getViewStats(
+                        MIN_TIME,
```
Дата старта - это самое раннее время публикации событий, среди переданных(переданного), так как до этого момента просмотров быть не может)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+    public static final LocalDateTime MAX_TIME = toTime("5000-01-01 00:00:00");
+    public static final LocalDateTime MIN_TIME = toTime("2000-01-01 00:00:00");
+
+    public static List<EventFullDto> getViews(List<EventFullDto> eventDtos, StatService statService) {
```
Мы не должно работать с бизнес-слоем в утилитарном классе) Вся бизнес логика должна находиться в сервисе

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+        return new ArrayList<>(views.values());
+    }
+
+    public static List<EventShortDto> getViewsToShort(List<EventShortDto> eventDtos, StatService statService) {
```
Мы не должно работать с бизнес-слоем в утилитарном классе) Вся бизнес логика должна находиться в сервисе

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public static void getConfirmedRequests(List<EventFullDto> eventDtos,
+                                            RequestsRepository requestsRepository) {
```
Аналогично и тут) В утилитарном классе мы не должны обращаться к репозиторию) Это взаимодействие должно осуществляться в сервисе)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public static void getConfirmedRequestsToShort(List<EventShortDto> eventDtos,
+                                                   RequestsRepository requestsRepository) {
```
Аналогично и тут) В утилитарном классе мы не должны обращаться к репозиторию) Это взаимодействие должно осуществляться в сервисе)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

GreNaz committed:
вынести версии зависимостей одной группы в блок properties

GreNaz committed:
вынести версии зависимостей одной группы в блок properties

GreNaz committed:
По спецификации это значение не является обязательным) Если оно не было передано - следует произвести получение без учета pinned

GreNaz committed:
Почему бы не использовать более декларативный подход и использовать аннотации ограничения) Например, для from можно было бы использовать @PositiveOrZero, так как значение может быть 0, либо больше) Аналогично и для size, но в этом случае следует использовать @Positive, так как значение не может быть равно 0

GreNaz committed:
Почему бы не использовать более декларативный подход и использовать аннотации ограничения) Например, для from можно было бы использовать @PositiveOrZero, так как значение может быть 0, либо больше) Аналогично и для size, но в этом случае следует использовать @Positive, так как значение не может быть равно 0

GreNaz committed:
Если параметр ids не передан - получение должно быть выполнено без его учета) Инициализация не требуется)

GreNaz committed:
Аннотация @Repository не требуется, так как наследуясь от JpaRepository мы уже объявляем бин)

GreNaz committed:
Действительно ли здесь требуется эта аннотация?)

GreNaz committed:
Можно было бы не дублировать хендлеры, а описать похожие в одном, используя @ExceptionHandler({@ExceptionHandler.class, ....})
В качестве параметра метода будет выступать ближайший родитель) В данном случае - это RuntimeException

GreNaz committed:
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем Throwable, а возвращать статус 500 INTERNAL_SERVER_ERROR
Так же следует добавить обработчик для исключения, которое выбрасывается при валидации данных с помощью аннотаций - MethodArgumentNotValidException и ConstraintViolationException, возвращать будет статус 400 BAD_REQUEST, так как ошибка связана с некорректными данными от пользователя)

GreNaz committed:
Используем exists, чтобы не производить лишнюю десериализацию) Но можно объявить свой метод deleteById, который не будет выбрасывать исключение)

GreNaz committed:
Аннотация также не требуется)

GreNaz committed:
Аннотация также не требуется)

GreNaz committed:
Аннотация также не требуется)

GreNaz committed:
Аннотация также не требуется)

GreNaz committed:
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода isBlank)
Иначе при обновлении допустимо передать строку только из пробелов или пустую, что при создании недопустимо)

GreNaz committed:
Лучше всегда указывать исключение, которое мы хотим выбросить и использовать то, что относится к этому виду ошибки) Чтобы из сообщения было проще понять, что именно пошло не так)

GreNaz committed:
Для начала следует проверить, что значение newCompilationDto.getEvents() не null, так как подборка может быть создана пустой)

GreNaz committed:
Убрал валидацию с ДТО

GreNaz committed:
В данном случае хорошо было бы использовать Set, так как события в подборке - уникальны)

GreNaz committed:
В данном случае хорошо было бы использовать Set, так как события в подборке - уникальны)

GreNaz committed:
Хранение вычисляемых полей в БД - не является хорошей практикой. Хорошо было бы исключить это поле из таблицы БД

Также в ТЗ явно прописано, что просмотры должны храниться в сервисе статистики)

GreNaz committed:
Хорошо было бы сделать тип - примитивом, так как значение по спецификации не является обязательным, а по дефолту должно быть не null, а false

GreNaz committed:
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться

GreNaz committed:
В данном случае хорошо было бы использовать не String, а Enum, чтобы заранее производить проверку а соответствие допустимых статусов)

GreNaz committed:
paid может иметь лишь два состояния) Хорошо было бы в данном случае использовать @NotNull, либо же изменить тип на примитив) Если значение не будет передано, в таком случае оно будет проинициализировано не null, а false)

GreNaz committed:
Аналогично, requestModeration может иметь лишь два состояния
По спецификации стандартным значением является true, можно было бы сразу же произвести инициализацию)

GreNaz committed:
Также хорошо было бы подправить тип на примитивный, так как @PositiveOrZero допускает null за корректное значение)

GreNaz committed:
Также хорошо было бы подправить тип на примитивный, так как @PositiveOrZero допускает null за корректное значение)

GreNaz committed:
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться
Также можно было бы в будущем классе dto навесить над соответствующими полями аннотации валидации, а здесь - @Valid, чтобы они стали учитываться) Так как сейчас допустимо передать пустой json

GreNaz committed:
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться
Также можно было бы в будущем классе dto навесить над соответствующими полями аннотации валидации, а здесь - @Valid, чтобы они стали учитываться) Так как сейчас допустимо передать пустой json

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,60 @@
+        Map<String, EventFullDto> views = eventDtos.stream().collect(Collectors.toMap(fullEventDto
+                -> "/events/" + fullEventDto.getId(), fullEventDto -> fullEventDto));
+        Object responseBody = statService.getViewStats(
+                        MIN_TIME,
```
Дата старта - это самое раннее время публикации событий, среди переданных(переданного), так как до этого момента просмотров быть не может)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,60 @@
+                -> "/events/" + fullEventDto.getId(), fullEventDto -> fullEventDto));
+        Object responseBody = statService.getViewStats(
+                        MIN_TIME,
+                        MAX_TIME,
```
Дата окончания - это текущая дата, так как позже нее не может быть просмотров)

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public void getConfirmedRequests(List<EventFullDto> eventDtos,
+                                     RequestsRepository requestsRepository) {
```
Утилитарный класс не должен работать с уровнем сервиса) Все бизнес-логика должна находиться в нем, в утилитарный класс мы должны лишь передавать те данные, над которыми следует произвести какую-то работу, так как работа с репозиторием и сервисом - это не область ответственности этого класса

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/repository/events/util/EventUtil.java:
```diff
@@ -0,0 +1,60 @@
+    public static final LocalDateTime MAX_TIME = LocalDateTime.parse("5000-01-01 00:00:00", FORMATTER);
+    public static final LocalDateTime MIN_TIME = LocalDateTime.parse("2000-01-01 00:00:00", FORMATTER);
+
+    public List<EventFullDto> getViews(List<EventFullDto> eventDtos, StatService statService) {
```
Утилитарный класс не должен работать с уровнем сервиса) Все бизнес-логика должна находиться в нем, в утилитарный класс мы должны лишь передавать те данные, над которыми следует произвести какую-то работу, так как работа с репозиторием и сервисом - это не область ответственности этого класса

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventUserRequest.java:
```diff
@@ -0,0 +1,39 @@
+    private String description; //  Новое описание
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate; //  Новые дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться
Здесь должен использоваться класс-DTO

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/model/events/dto/UpdateEventAdminRequest.java:
```diff
@@ -0,0 +1,39 @@
+    private String description; //  Новое описание
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate; //  Новые дата и время на которые намечено событие. Дата и время указываются в формате "yyyy-MM-dd HH:mm:ss"
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться
Здесь должен использоваться класс-DTO

TyutterinYakov commented on ewm-main/src/main/java/ru/practicum/ewm/service/events/impl/EventServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+        List<EventFullDto> fullEventDtoList = events.stream()
+                .map(EventMapper.EVENT_MAPPER::toEventFullDto)
+                .collect(Collectors.toList());
+        fullEventDtoList.forEach(event -> event.setConfirmedRequests(requestsRepository.findByEventIdConfirmed(event.getId()).size()));
```
текущей реализации мы плодим n-лишние запросы, так как производим обращение к методу репозитория - в цикле) Можно было бы объявить поле `confirmedRequests` в классе `Event`, навесив над ним `@Transient`, а затем просетить всем полученным событиям - это значение, предварительно получив их разом из БД) Либо же можно было бы не вводить поле, а просто получив бы количество подтвержденных заявок в виде мапы, а затем на основе этих значений - уже фильтровать события) Можно воспользоваться и описанным на прошлом ревью механизмом проекций, но тут по желанию) Главное уйти от n-лишних запросов)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил оставшиеся недочеты, которые следует подправить) Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

GreNaz committed:
Дата старта - это самое раннее время публикации событий, среди переданных(переданного), так как до этого момента просмотров быть не может)
Дата окончания - это текущая дата, так как позже нее не может быть просмотров)

GreNaz committed:
Утилитарный класс не должен работать с уровнем сервиса) Все бизнес-логика должна находиться в нем, в утилитарный класс мы должны лишь передавать те данные, над которыми следует произвести какую-то работу, так как работа с репозиторием и сервисом - это не область ответственности этого класса

GreNaz committed:
текущей реализации мы плодим n-лишние запросы, так как производим обращение к методу репозитория - в цикле) Можно было бы объявить поле confirmedRequests в классе Event, навесив над ним @Transient, а затем просетить всем полученным событиям - это значение, предварительно получив их разом из БД) Либо же можно было бы не вводить поле, а просто получив бы количество подтвержденных заявок в виде мапы, а затем на основе этих значений - уже фильтровать события) Можно воспользоваться и описанным на прошлом ревью механизмом проекций, но тут по желанию) Главное уйти от n-лишних запросов)

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта! Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного вечера!))
